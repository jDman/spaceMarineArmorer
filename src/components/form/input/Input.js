import React from 'react';

import classes from './Input.module.css';

const Input = props => {
  let inputElement;
  let errorElement;

  const inputClasses = [classes.InputElement];

  if (props.invalid && props.hasValidation && props.touched) {
    inputClasses.push(classes.Invalid);
    errorElement = <p className={classes.Error}>Please enter a valid value!</p>;
  }

  switch (props.elementType) {
    case 'input':
      inputElement = (
        <input
          className={inputClasses.join(' ')}
          {...props.elementConfig}
          value={props.value}
          onChange={props.changed}
        />
      );
      break;
    case 'textarea':
      inputElement = (
        <textarea
          className={inputClasses.join(' ')}
          {...props.elementConfig}
          value={props.value}
          onChange={props.changed}
        ></textarea>
      );
      break;
    case 'select':
      inputElement = (
        <select
          className={inputClasses.join(' ')}
          type={props.elementConfig.selectAttributes.type}
          name={props.elementConfig.selectAttributes.name}
          required={props.elementConfig.selectAttributes.required}
          value={props.value}
          onChange={props.changed}
        >
          {props.elementConfig.options.map(option => (
            <option key={option.value} value={option.value}>
              {option.displayValue}
            </option>
          ))}
        </select>
      );
      break;
    default:
      inputElement = (
        <input
          className={inputClasses.join(' ')}
          {...props.elementConfig}
          value={props.value}
          onChange={props.changed}
        />
      );
  }

  return (
    <div className={classes.Input}>
      <label className={classes.Label}>{props.label}</label>
      {inputElement}
      {errorElement}
    </div>
  );
};

export default Input;
