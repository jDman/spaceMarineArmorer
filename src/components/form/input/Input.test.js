import React from 'react';
import renderer from 'react-test-renderer';
import Input from './Input';

describe('<Input />', () => {
  it('renders Input component with props passed in for elementType input text', () => {
    const config = {
      elementType: 'input',
      elementConfig: {
        type: 'text',
        name: 'testText',
        placeholder: 'This test is...',
        required: true
      },
      value: '',
      valid: false,
      touched: false,
      validation: { required: true }
    };
    const component = renderer.create(
      <Input
        elementType={config.elementType}
        elementConfig={config.elementConfig}
        value={config.value}
        invalid={!config.valid}
        touched={config.touched}
        hasValidation={config.validation}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Input component with props passed in for elementType input number', () => {
    const config = {
      elementType: 'input',
      elementConfig: {
        type: 'number',
        name: 'testNumber',
        required: true,
        min: 0,
        max: 10000,
        step: 100
      },
      value: 0,
      valid: false,
      touched: false,
      validation: { required: true }
    };
    const component = renderer.create(
      <Input
        elementType={config.elementType}
        elementConfig={config.elementConfig}
        value={config.value}
        invalid={!config.valid}
        touched={config.touched}
        hasValidation={config.validation}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Input component with props passed in for elementType textarea', () => {
    const config = {
      elementType: 'textarea',
      elementConfig: {
        name: 'testTextarea',
        placeholder: 'This test is...',
        required: true,
        minLength: 5,
        maxLength: 2000
      },
      value: '',
      validation: {
        required: true,
        minLength: 5,
        maxLength: 2000
      },
      value: '',
      valid: false,
      touched: false
    };
    const component = renderer.create(
      <Input
        elementType={config.elementType}
        elementConfig={config.elementConfig}
        value={config.value}
        invalid={!config.valid}
        touched={config.touched}
        hasValidation={config.validation}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Input component with props passed in for elementType select', () => {
    const config = {
      elementType: 'select',
      elementConfig: {
        selectAttributes: {
          type: 'select',
          name: 'testSelect',
          required: true
        },
        options: [
          { value: 'test1', displayValue: 'Test 1' },
          { value: 'test2', displayValue: 'Test 2' },
          { value: 'test3', displayValue: 'Test 3' }
        ]
      },
      value: '',
      validation: {
        required: true
      },
      valid: false,
      touched: false
    };
    const component = renderer.create(
      <Input
        elementType={config.elementType}
        elementConfig={config.elementConfig}
        value={config.value}
        invalid={!config.valid}
        touched={config.touched}
        hasValidation={config.validation}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
