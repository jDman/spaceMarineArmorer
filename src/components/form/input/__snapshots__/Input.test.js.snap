// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Input component 1`] = `
<div
  className="Input"
>
  <label
    className="Label"
  />
  <input
    className="InputElement"
  />
</div>
`;

exports[`renders Input component with props passed in for elementType input number 1`] = `
<div
  className="Input"
>
  <label
    className="Label"
  />
  <input
    className="InputElement"
    max={10000}
    min={0}
    name="testNumber"
    required={true}
    step={100}
    type="number"
    value={0}
  />
</div>
`;

exports[`renders Input component with props passed in for elementType input text 1`] = `
<div
  className="Input"
>
  <label
    className="Label"
  />
  <input
    className="InputElement"
    name="testText"
    placeholder="This test is..."
    required={true}
    type="text"
    value=""
  />
</div>
`;

exports[`renders Input component with props passed in for elementType select 1`] = `
<div
  className="Input"
>
  <label
    className="Label"
  />
  <select
    className="InputElement"
    name="testSelect"
    required={true}
    type="select"
    value=""
  >
    <option
      value="test1"
    >
      Test 1
    </option>
    <option
      value="test2"
    >
      Test 2
    </option>
    <option
      value="test3"
    >
      Test 3
    </option>
  </select>
</div>
`;

exports[`renders Input component with props passed in for elementType textarea 1`] = `
<div
  className="Input"
>
  <label
    className="Label"
  />
  <textarea
    className="InputElement"
    maxLength={2000}
    minLength={5}
    name="testTextarea"
    placeholder="This test is..."
    required={true}
    value=""
  />
</div>
`;
